server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -84,10 +111,32 @@ MappedFieldType get(String field) {
         if (fieldType != null) {
             return fieldType;
         }
+        return getDynamicField(field);
+    }
+
+    private MappedFieldType getDynamicField(String field) {
```

Comment:
Was there a motivation for changing the lookup approach? The old approach seemed more streamlined -- this one seems to do several passes through the string (in `longestPossibleParent`, then the `contains` and `lastIndexOf` calls in a loop below).

Reply:
I'm not sure why I re-implemented this, but the original approach is clearly more efficient.  I've updated.